src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs

```diff
@@ -678,6 +704,84 @@ public void TestMergeChanges_Overlapping_OldBeforeNew()
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 3), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 0), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa4", change1.ToString());
+            Assert.Equal("0bbaa4", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 3), "bbaa") }, changes);
+        }
+
+        [Fact]
+        public void TestMergeChanges_SameStart_AndNewHasDeletion_SmallerThanOldInsertion()
+        {
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 0), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 1), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa1234", change1.ToString());
+            Assert.Equal("0bba1234", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 0), "bba") }, changes);
+        }
+
+        [Fact]
+        public void TestMergeChanges_SameStart_AndNewHasDeletion_EqualToOldInsertion()
+        {
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 0), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 2), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa1234", change1.ToString());
+            Assert.Equal("0bb1234", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 0), "bb") }, changes);
+        }
+
+        [Fact]
+        public void TestMergeChanges_SameStart_AndNewHasDeletion_LargerThanOldInsertion()
+        {
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 0), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 3), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa1234", change1.ToString());
+            Assert.Equal("0bb234", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 1), "bb") }, changes);
+        }
+
+        [Fact]
+        public void TestMergeChanges_SameStart_AndBothHaveDeletion_NewDeletionSmallerThanOld()
+        {
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 3), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 1), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa4", change1.ToString());
+            Assert.Equal("0bba4", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 3), "bba") }, changes);
+        }
```

Comment:
Are these all cases that would have thrown before your change?

BTW, this new form for the tests is significantly easier to reason about, thanks.

Reply:
Only 3 of them would throw, but I no longer remember which ones. `TestMergeChanges_Overlapping_NewInsideOld_AndOldHasDeletion` for sure. We can revert the class itself on local and run tests to see.

Reply:
I ended up pulling down the changes after I commented to see what would break, and it was pretty much as you describe :)

I actually needed to pull down the changes and run the whole test class with the debugger with breakpoints on areas you changed to be sure I understood what the code was doing.

Reply:
Thanks for your time! Yep it was a bit hard to follow for me as well, existing tests were extremely useful.